File: backend/api/utils/audit.py
"""
Audit logging utility for TerraCrate.

Provides a single ``log_audit()`` helper that writes to the ``audit_logs``
table.  Failures are silently caught so that audit logging never breaks the
main request flow.
"""

import json
import logging

from flask import has_request_context, request

from models import AuditLog, db

logger = logging.getLogger(__name__)


def log_audit(
    action: str,
    *,
    target_type: str | None = None,
    target_id: str | None = None,
    description: str | None = None,
    status: str = "success",
    metadata: dict | None = None,
    user_id: int | None = None,
    user_email: str | None = None,
):
    """Record an audit log entry.

    Parameters
    ----------
    action : str
        Dotted action identifier, e.g. ``"auth.login"``, ``"file.upload"``.
    target_type : str, optional
        Kind of entity acted upon (``"user"``, ``"file"``, ``"group"``, …).
    target_id : str, optional
        Identifier of the target (user id, file path, group id, …).
    description : str, optional
        Human-readable summary of what happened.
    status : str
        ``"success"`` (default) or ``"failure"``.
    metadata : dict, optional
        Extra context serialised as JSON (old/new values, sizes, …).
    user_id : int, optional
        Override – pulled from ``request.user`` when omitted.
    user_email : str, optional
        Override – pulled from ``request.user`` when omitted.
    """
    try:
        # Auto-detect user context from Flask request
        if has_request_context():
            req_user = getattr(request, "user", None)
            if req_user and user_id is None:
                user_id = req_user.get("user_id")
            if req_user and user_email is None:
                user_email = req_user.get("email")

        # Capture IP address
        ip_address = None
        if has_request_context():
            ip_address = request.headers.get("X-Forwarded-For", request.remote_addr)

        entry = AuditLog(
            user_id=user_id,
            user_email=user_email,
            action=action,
            target_type=target_type,
            target_id=str(target_id) if target_id is not None else None,
            description=description,
            ip_address=ip_address,
            status=status,
            metadata_json=json.dumps(metadata) if metadata else None,
        )
        db.session.add(entry)
        db.session.commit()
    except Exception:
        logger.debug("Failed to write audit log entry", exc_info=True)
        try:
            db.session.rollback()
        except Exception:
            pass

File: backend/internal/audit/audit.go
// Package audit writes best-effort entries to the audit_logs table; a
// failure to log never breaks the request flow (matching utils/audit.py).
package audit

import (
	"log/slog"

	"github.com/andreas-04/terra-crate/backend/internal/store"
)

type Logger struct {
	store *store.Store
}

func New(s *store.Store) *Logger { return &Logger{store: s} }

// Entry captures the optional context for one audit record.
type Entry struct {
	TargetType  string
	TargetID    string
	Description string
	Status      string // "" defaults to "success"
	UserID      int    // 0 = unknown
	UserEmail   string
	IPAddress   string
}

// Log records an action; errors are swallowed (logged at debug level only).
func (l *Logger) Log(action string, e Entry) {
	rec := &store.AuditLog{Action: action, Status: e.Status}
	if e.TargetType != "" {
		rec.TargetType = &e.TargetType
	}
	if e.TargetID != "" {
		rec.TargetID = &e.TargetID
	}
	if e.Description != "" {
		rec.Description = &e.Description
	}
	if e.UserID != 0 {
		rec.UserID = &e.UserID
	}
	if e.UserEmail != "" {
		rec.UserEmail = &e.UserEmail
	}
	if e.IPAddress != "" {
		rec.IPAddress = &e.IPAddress
	}
	if err := l.store.InsertAuditLog(rec); err != nil {
		slog.Warn("audit: failed to write log entry", "error", err)
	}
}
